### PR TITLE
[PF-2833] Add GCE instance resource type

### PR DESCRIPTION
**Pull request by iliadsh**
Sorry for the large delta, but it's mostly just boilerplate.

### CODE REVIEW

**Review by ddietterich — COMMENTED** [review]
Nice work. The main issue is getting the resource attributes right.

I think it is worth taking a second pass over the code to factor out common parts that are shared by notebook and gce. I found a few, but likely missed some.

Also, please add to `bio/terra/workspace/service/resource/statetests/GcpResourceStateFailureTest.java` to validate that the state checks are working for the new endpoints.

File: openapi/src/parts/controlled_gcp_gce_instance.yaml [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+      type: object
+      required: [ machineType ]
+      properties:
+        instanceId:
+          description: >-
+            An instanceId unique to this project and location. Must be 1-63 characters, using lower
+            case letters, numbers, and dashes. The first character must be a lower case letter,
+            and the last character must not be a dash.
+          type: string
+        zone:
```
[the remark this answers is not in the document]
Comment (iliadsh) [reply openapi/src/parts/controlled_gcp_gce_instance.yaml:190]: I'm not sure if I'm missing something obvious but I couldn't find a region in any of the common creation params. FWIW all other controlled gcp resources pass in a location.

File: openapi/src/parts/controlled_gcp_gce_instance.yaml [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        guestAccelerators:
+          type: array
+          items:
+            $ref: '#/components/schemas/GcpGceInstanceGuestAccelerator'
+        metadata:
+          description: >-
+            Custom metadata to apply to this instance. Some metadata keys are reserved by Terra and
+            will throw a bad request error if specified: \'terra-workspace-id\', \'terra-cli-server\'.
+          type: object
+          additionalProperties:
```
[the remark this answers is not in the document]
Comment (iliadsh) [reply openapi/src/parts/controlled_gcp_gce_instance.yaml:206]: `additionalProperties` turns this into a Map<String, String> https://swagger.io/docs/specification/data-models/dictionaries/

File: openapi/src/resources/gcp_gce_instance.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+components:
+  schemas:
+    GcpGceInstanceAttributes:
+      description: >-
+        GCE instance properties included in post-creation, get, and update. Others must be
+        retrieved from GCP using the specification.
+      type: object
+      required: [ projectId, zone, instanceId ]
+      properties:
```
[the remark this answers is not in the document]
Comment (iliadsh) [reply openapi/src/resources/gcp_gce_instance.yaml:9]: Sg, though it feels weird using 'region' for all the zone APIs (even if we store a zone in there)

File: openapi/src/parts/controlled_gcp_gce_instance.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (ddietterich) [reply openapi/src/parts/controlled_gcp_gce_instance.yaml:190]: Yes, you are right. I am confusing the internal structures with the REST API.

File: openapi/src/parts/controlled_gcp_gce_instance.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ddietterich) [reply openapi/src/parts/controlled_gcp_gce_instance.yaml:206]: Huh! TIL

File: openapi/src/resources/gcp_gce_instance.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddietterich) [reply openapi/src/resources/gcp_gce_instance.yaml:9]: Yes, I retract my comment.

**Review by yuhuyoyo — COMMENTED** [review]
what's the bigger context for this? I must have missed some docs

File: service/src/test/java/bio/terra/workspace/common/fixtures/ControlledGcpResourceFixtures.java [hunk 4]
```diff
@@ -347,4 +350,57 @@ public static ControlledAiNotebookInstanceResource.Builder makeDefaultAiNotebook
...
+        .managedBy(ManagedByType.MANAGED_BY_USER)
+        .createdByEmail(MockMvcUtils.DEFAULT_USER_EMAIL)
+        .region(DEFAULT_RESOURCE_REGION);
+  }
+
+  public static ControlledGceInstanceResource.Builder makeDefaultGceInstance() {
+    return ControlledGceInstanceResource.builder()
+        .common(makeGceInstanceCommonFieldsBuilder().build())
+        .instanceId(TestUtils.appendRandomNumber("my-cloud-id"))
+        .zone("us-east1-b")
```
Comment (yuhuyoyo) [inline service/src/test/java/bio/terra/workspace/common/fixtures/ControlledGcpResourceFixtures.java:395]: is DEFAULT_RESOURCE_REGION at line 388 us-east1?

**Comment by iliadsh** [comment]
> what's the bigger context for this? I must have missed some docs

My understanding is that while we flesh out requirements and overall design for apps, adding instance support to WSM would be a largely uncontroversial change that unblocks some other work like prototyping an end-to-end RStudio startup.

File: service/src/test/java/bio/terra/workspace/common/fixtures/ControlledGcpResourceFixtures.java [hunk 5]
```diff
@@ -347,4 +350,57 @@ public static ControlledAiNotebookInstanceResource.Builder makeDefaultAiNotebook
...
+        .projectId("my-project-id");
+  }
+
+  public static ControlledGceInstanceResource.Builder makeDefaultGceInstance(UUID workspaceId) {
```
Comment (yuhuyoyo) [inline service/src/test/java/bio/terra/workspace/common/fixtures/ControlledGcpResourceFixtures.java:399]: is this method used?

File: service/src/test/java/bio/terra/workspace/common/fixtures/ControlledGcpResourceFixtures.java [hunk 6]
```diff
@@ -347,4 +350,57 @@ public static ControlledAiNotebookInstanceResource.Builder makeDefaultAiNotebook
...
+      new ApiGcpGceUpdateParameters()
+          .metadata(ImmutableMap.of("sky", "blue", "rose", "red", "foo", "bar2", "count", "0"));
+
+  public static final ApiGcpGceUpdateParameters GCE_INSTANCE_UPDATE_PARAMETERS =
+      new ApiGcpGceUpdateParameters().metadata(ImmutableMap.of("foo", "bar", "count", "3"));
+
+  public static ApiGcpGceInstanceCreationParameters defaultGceInstanceCreationParameters() {
+    return new ApiGcpGceInstanceCreationParameters()
+        .instanceId(TestUtils.appendRandomNumber("default-instance-id"))
+        .zone("us-central1-a")
```
Comment (yuhuyoyo) [inline service/src/test/java/bio/terra/workspace/common/fixtures/ControlledGcpResourceFixtures.java:366]: consider extract this zone to a static field to be shared.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/ControlledAiNotebookInstanceResource.java [hunk 7]
```diff
@@ -159,43 +159,6 @@ public Optional<UniquenessCheckAttributes> getUniquenessCheckAttributes() {
             .addParameter("instanceId", getInstanceId()));
   }
 
-  /**
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/ControlledAiNotebookInstanceResource.java:162]: is this related?

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/UpdateGceInstanceResourceZoneAttributesStep.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import bio.terra.workspace.db.DbSerDes;
+import bio.terra.workspace.db.ResourceDao;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys.ResourceKeys;
+
+public class UpdateGceInstanceResourceZoneAttributesStep implements Step {
+
+  private final ControlledGceInstanceResource resource;
+  private final ResourceDao resourceDao;
+
+  public UpdateGceInstanceResourceZoneAttributesStep(
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/UpdateGceInstanceResourceZoneAttributesStep.java:19]: this look literally the same as updateNotebookResourceLocationAttributesStep. Can we pass in the resource type in the constructor to know which attributes to use and merge the two classes?

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/RetrieveNetworkNameStep.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    }
+    pickAndStoreNetwork(subnetworks, flightContext.getWorkingMap());
+    return StepResult.getStepResultSuccess();
+  }
+
+  /**
+   * Fetches the valid zone given resource "zone", since the default workspace region might be
+   * passed in if no zone is specified. If none is found, returns the resource zone attributes.
+   */
+  private String maybeGetValidZone(String projectId) throws IOException {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/RetrieveNetworkNameStep.java:68]: same with this step, it looks almost the same. please merge.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/GceInstanceCloudSyncStep.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+  /**
+   * Creates the bindings to set on the compute instance.
+   *
+   * <p>{@link
+   * bio.terra.workspace.service.resource.controlled.ControlledResourceService#configureGcpPolicyForResource}
+   * works in com.google.cloud.Policy objects, but these are not used by the compute engine API.
+   * Transform the com.google.cloud.Policy into a list of bindings to use for the GCP GCE API.
+   */
+  private List<Binding> createBindings(GcpCloudContext cloudContext, FlightMap workingMap)
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/GceInstanceCloudSyncStep.java:94]: merge this method and the one below to be shared.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+import org.springframework.http.HttpStatus;
+
+/**
+ * A step for creating the GCE instance in the Google cloud.
+ *
+ * <p>Undo deletes the created instance.
+ */
+public class CreateGceInstanceStep implements Step {
+  /** Service account for the instance needs to contain these scopes to interact with SAM. */
+  private static final List<String> SERVICE_ACCOUNT_SCOPES =
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java:61]: share this field

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    instance.setMetadata(
+        new Metadata()
+            .setItems(
+                metadata.entrySet().stream()
+                    .map(i -> new Items().setKey(i.getKey()).setValue(i.getValue()))
+                    .collect(Collectors.toList())));
+    return instance;
+  }
+
+  private static void addDefaultMetadata(
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java:223]: share this.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+    String subnetworkName = workingMap.get(CREATE_GCE_INSTANCE_SUBNETWORK_NAME, String.class);
+    instance.setNetworkInterfaces(
+        List.of(
+            new NetworkInterface()
+                .setNetwork(String.format("projects/%s/global/networks/%s", projectId, networkName))
+                .setSubnetwork(
+                    String.format(
+                        "projects/%s/regions/%s/subnetworks/%s", projectId, region, subnetworkName))
+                .setAccessConfigs(
+                    List.of(new AccessConfig().setType("ONE_TO_ONE_NAT").setName("External IP")))));
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java:250]: extract private static final field for ONE_TO_ONE_NAT and add a description. same for External IP.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+  }
+
+  private static void setNetworks(Instance instance, String projectId, FlightMap workingMap) {
+    String region = workingMap.get(CREATE_RESOURCE_REGION, String.class);
+    String networkName = workingMap.get(CREATE_GCE_INSTANCE_NETWORK_NAME, String.class);
+    String subnetworkName = workingMap.get(CREATE_GCE_INSTANCE_SUBNETWORK_NAME, String.class);
+    instance.setNetworkInterfaces(
+        List.of(
+            new NetworkInterface()
+                .setNetwork(String.format("projects/%s/global/networks/%s", projectId, networkName))
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java:245]: have a share method for constructing network and subnetowrk string.

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+                .setSubnetwork(
+                    String.format(
+                        "projects/%s/regions/%s/subnetworks/%s", projectId, region, subnetworkName))
+                .setAccessConfigs(
+                    List.of(new AccessConfig().setType("ONE_TO_ONE_NAT").setName("External IP")))));
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    final GcpCloudContext gcpCloudContext =
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java:255]: rm all the final on local vars.

File: service/src/test/java/bio/terra/workspace/common/fixtures/ControlledGcpResourceFixtures.java (hunk 4 above)
Comment (iliadsh) [reply service/src/test/java/bio/terra/workspace/common/fixtures/ControlledGcpResourceFixtures.java]: it's us-central1, but I don't think it's used for anything here

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/ControlledAiNotebookInstanceResource.java (hunk 7 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/ainotebook/ControlledAiNotebookInstanceResource.java:162]: Since we do a very similar cloud sync I assumed we would have the same DRS slowness

File: service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java (hunk 12 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/gceinstance/CreateGceInstanceStep.java:221]: The notebooks also have some proxy specific metadata and it's unclear to me if the GCP instances will need some sort of equivalent yet

**Review by yuhuyoyo — APPROVED** [review]

**Comment by iliadsh** [comment]
Thanks for the reviews!
